### PR TITLE
feat: codegen go-bindnode support

### PR DESCRIPTION
**Pull request by aschmahmann**

### CODE REVIEW

File: app/basic/put.go
Comment (aschmahmann) [inline app/basic/put.go]: @warpfork not sure how to fix this, but can rebase on master once go-ipldtool is bumped.

File: app/schema/schema.go
Comment (aschmahmann) [inline app/schema/schema.go]: @mvdan @warpfork not sure how happy we are with this approach. We could also just let users figure it out themselves or copy the thing we're embedding.

File: app/schema/schema.go
Comment (aschmahmann) [inline app/schema/schema.go:183]: @mvdan AFAICT this function is currently broken even for simple schemas since the Link type is unspecified.

File: app/schema/schema.go
Comment (aschmahmann) [inline app/schema/schema.go:58]: I feel like we should have some tests here that show at least basic compilation of the generated code.

@warpfork any ideas/suggestions? If you're able to put together one for go-gengo it'd probably be easy to adapt to go-bindnode.

Comment (warpfork) [reply app/schema/schema.go:58]: Yeah, been wondering that to myself too.  Do you think a test that just lists that some files come out is enough?

Maybe we extend that to also do `head -n 10 {output}/*.go` ?

I don't think it's viable to emit whole bodies and test against that.  For the go-bindnode version and small schemas, that might actually be tractable volumes of text, but for go-gengo, it's definitely not.

File: app/basic/put.go
Comment (warpfork) [reply app/basic/put.go]: I was able to fix this by bumping the version of the dsadapter module.  I pushed a commit to your branch with that -- hope that's okay.  Squash freely.

Comment (aschmahmann) [reply app/basic/put.go]: Yes, thank you 🙏

File: app/schema/schema.go
Comment (aschmahmann) [reply app/schema/schema.go:58]: > I don't think it's viable to emit whole bodies and test against that

What part seems unreasonable/a pain here? I'm thinking that some basic "does it compile" and "does `init` run" tests are probably all we need for now.

Most of the testing should probably live in go-ipld-prime next to the generation code, or perhaps as a separate package that runs `go generate` because Go makes linking in code dynamically pretty difficult.

File: app/schema/schema.go
Comment (mvdan) [inline app/schema/schema.go]: just use filepath.Join?

File: app/schema/schema.go
Comment (mvdan) [reply app/schema/schema.go]: FYI, ditto about filepath.Join.

File: app/schema/schema.go
Comment (mvdan) [inline app/schema/schema.go]: or just use `[]byte` to avoid the conversion below

File: app/schema/schema.go
Comment (mvdan) [inline app/schema/schema.go]: how about just `panic(err)`? you should include the error string somehow, at least.

File: app/schema/schema.go
Comment (mvdan) [inline app/schema/schema.go]: I think exposing the entire TypeSystem is not something we should encourage, as it makes the API surface too large. And since types are looked up by string, it's easy to make typos and the compiler won't help you against them.

I think exposing the schema.TypedPrototype directly is best, e.g. https://github.com/ipld/go-ipld-prime/blob/c0e475c07685f9f7f64e7e34bb9ffe9c93b5e3ee/schema/dmt/schema.go#L15-L17.

There I followed the struct approach like with gengo's generated API, so that we can easily fit more schemas if needed. But I could have just as easily used `var Prototype schema.TypedPrototype`.

File: app/schema/schema.go
Comment (mvdan) [inline app/schema/schema.go]: Hmm, IMO we should encourage the user to put their ipldsch schema file in the same directory where it's embedded, parsed, and used. Otherwise you can get into weird situations such as "the schema file is outside the module so it's excluded from module zips", or "the schema file is outside the package so it was excluded from `go mod vendor`".

One way to do this would be to copy the schema file in the output directory. The other way would be to expect the schema file to already exist in the output directory, and error otherwise. I think I'd lean towards the former, with a printed warning that we've made a copy of the schema file.

File: app/schema/schema.go
Comment (aschmahmann) [reply app/schema/schema.go]: > I think exposing the schema.TypedPrototype directly is best, e.g. https://github.com/ipld/go-ipld-prime/blob/c0e475c07685f9f7f64e7e34bb9ffe9c93b5e3ee/schema/dmt/schema.go#L15-L17.

If we're going to do this it probably requires generating a file of a similar layout to gengo's `ipldsch_types.go` where there's a single `var Schemas schemaSlab` where `schemaSlab` is a struct that has fields named after each IPLD schema type and of type `schema.TypedPrototype`.

We can prototype this here, but my guess is it should probably live in go-ipld-prime

File: app/schema/schema.go
Comment (aschmahmann) [reply app/schema/schema.go]: Copy doesn't seem so bad, but maybe it's worth adding a flag to opt-out.

I'm thinking that I could see a user wanting a directory layout like:
- ipld-schemas
   - the-schemas
   - generated
      - doc.go with a `go generate` command
      - gengo
      - bindnode
      
But maybe using multiple versions of codegen for a single language together is sort of a crazy thing

File: app/schema/schema.go
Comment (aschmahmann) [reply app/schema/schema.go]: Updated, lmk if this is what you had in mind

File: app/schema/schema.go
Comment (aschmahmann) [inline app/schema/schema.go]: Hack so things can work in the meanwhile 🙈

File: app/schema/schema.go
Comment (aschmahmann) [inline app/schema/schema.go]: Is doing `5:` reliable here? Is there a way for me to isolate the primitive types without hard coding the words "Bool", "Int", ...?

Comment (aschmahmann) [reply app/schema/schema.go]: Similarly, it appears that the bindnode codegen skips over type aliases (e.g. `type Error string`). Is this intentional and if so is there a similar way for me to check if a type is primitive without hard coding all the primitives?

It's not so much and hopefully we won't be changing what primitives are called or how many there are, but figured I'd ask if these functions were already around.

File: app/schema/schema.go
Comment (warpfork) [reply app/schema/schema.go:58]: hm.  Yes, we could do `ls` as a test to make sure the expected files come out at all, and then `go test .` to see if it compiles, I suppose.  That's brief and covers a lot of ground.

(By "test against whole bodies" I mean, like, embedding a whole copy of a codegen output run as a fixture.  That'd be too voluminous and fairly silly.)

File: app/schema/schema.go
Comment (mvdan) [reply app/schema/schema.go:183]: That oversight is now fixed as per https://github.com/ipld/go-ipld-prime/pull/285

File: app/schema/schema.go
Comment (mvdan) [reply app/schema/schema.go]: My intuition is that basically noone will want to use both gengo and bindnode codegen at the same time. Users will want to choose one or the other. At most, they might switch from one to the other, if they find they chose the wrong one given their tradeoffs/priorities.

File: app/schema/schema.go
Comment (mvdan) [reply app/schema/schema.go]: If this was for Link, you can probably remove this?

File: app/schema/schema.go
Comment (mvdan) [reply app/schema/schema.go]: bindnode codegen indeed skips over "prelude" types, because they are handled with already-existing types. I personally think gengo should do the same - by using basicnode implementations and such.

As for the `5`, I think what you presumably want is the length of the prelude. We shouldn't really be hard-coding that. But I think we should more easily fix it by making gengo not generate code for prelude types.

Comment (mvdan) [reply app/schema/schema.go]: If gengo does need to generate code for prelude types, then I think the answer is that we should let gengo generate for those. I don't really know gengo closely enough to tell for sure. @warpfork will know.

File: app/schema/schema.go
Comment (warpfork) [inline app/schema/schema.go]: This doesn't appear to do quite what it says on the tin -- it's called `Types` but all the fields are `TypedPrototype` (so, prototypes, not types).

(I realize you're probably riffing on something the gengo package does, which has the same mismatch... gengo is wrong on this too, and that bit should be revisited there.)

File: app/schema/schema.go
Comment (warpfork) [inline app/schema/schema.go]: Possibly this bit about having a struct full of convenient access to prototypes should be optional?  Bindnode can be used without it.  (Gengo has a table like this, but _can't_ be used without it.)

Or if we do have this table, perhaps it should be made by the upstream bindnode gen code?

File: app/schema/schema.go
Comment (warpfork) [reply app/schema/schema.go]: Yeah, I would hope that no action is needed here, and the generator should own the choice of not generating any content that it doesn't need to.

(Sidenote, I softcore think the bindnode generator should emit `type String string` aliases, derpy though that may look in golang, just for completeness and consistency of rules, because it's essentially free (afaik), aside from looking funny.  But that's tangential to whether anything gets filtered up here, and also a weakly held opinion.)

File: app/schema/schema.go
Comment (warpfork) [reply app/schema/schema.go]: Copy plus opt-out seems fine.

May be worth mentioning here: @rvagg has at some point produced a tool which would let you have your IPLD Schema interspersed in a markdown file (or multiple markdown files) by just plucking out all code blocks tagged "`ipldsch`".  I don't think we have that capability in our golang code today.  But I have to say, I thought the idea was rather cool and ideally would like to keep that possibility floating around and relatively accessible.  It's very nice for making specifications and documentation converge.

**Comment by petar** [comment]
@aschmahmann Skimming through the comments, it doesn't look like merging this PR is a small project. Am I missing something?

**Comment by aschmahmann** [comment]
@petar AFAICT the remaining items are:
 - [ ] Remove the hack with "Bogus" in it (i.e. the last commit) and update to the latest go-ipld-prime
 - [ ] A rename @warpfork requested
 - [ ] A flag to make the prototypes optional (I'd make it a separate commit so if it comes out too ugly we can just drop it)
 - [ ] A flag to enable copying the schema file into the output that defaults to true
 - [ ] Leave the `5:` hack in there until @warpfork or @mvdan come up with a better idea 😄
 
 So basically two optional flags around emitting data we already have.

File: app/schema/schema.go
Comment (petar) [reply app/schema/schema.go]: As far as I can tell, opting out is not possible: If the idea of "opt-out" is to go:embed the original .ipldsch file (instead of the copied one), then this requires a go:embed directive with a relative path (like go:embed ../some.ipldsch). This is not supported by go:embed. Further, I don't see reasons for the opt out feature: The output directory is fully dedicated to (and owned by) code generation, and the user should have no preferences as to what goes there. Having choices only complicates their mental model.

**Comment by petar** [comment]
@aschmahmann:

> @petar AFAICT the remaining items are:
> 
> * [x]  Remove the hack with "Bogus" in it (i.e. the last commit) and update to the latest go-ipld-prime
Done.

> * [ ]  A rename @warpfork requested
Not sure what you are referring to? Renamed "Type" to "Prototypes"?

> * [ ]  A flag to enable copying the schema file into the output that defaults to true
See my comment in the thread. I don't think this is well justified.

> * [x]  Leave the `5:` hack in there until @warpfork or @mvdan come up with a better idea 😄
I had to use `6:` (after dep upgrade) to skip generation of `Link` type.

**Comment by petar** [comment]
@aschmahmann @warpfork To summarize what's left open on this one:
I don't think generation of prototypes should be optional. I think they should be either in or out. It is bad UX to have  optionality in codegen. (For comparison, FileDescriptors are included in protobuf, no options are provided.) Keeping them in requires copying the schema file to the output directory. I don't see any problems with that either. (Also see my comments in the thread.)

File: app/schema/schema.go
Comment (aschmahmann) [reply app/schema/schema.go]: @petar this is the rename referenced https://github.com/ipld/go-ipldtool/pull/3#issuecomment-1004331185

Comment (petar) [reply app/schema/schema.go]: Done.

File: app/schema/schema.go
Comment (aschmahmann) [reply app/schema/schema.go]: I don't have particularly strong feelings here. It seems like "copy the files around with go:generate" is the pattern https://github.com/golang/go/issues/46056, although perhaps @mvdan or @warpfork have other ideas.

As long as the user can reasonably have a single place to edit their schema no matter which/how many codegens they're using then SGTM.

**Comment by mvdan** [comment]
I agree with @petar that we shouldn't have too many knobs here.

I realise that there are hacks still present, and that we might need to tweak how the codegen works as we experiment with it. I don't think we should hold up the merge here. We can improve over time with more PRs :)

**Comment by mvdan** [comment]
But also: could you please open short reminder issues on https://github.com/ipld/go-ipld-prime with the API quirks that should be fixed, like that `5:` subslicing when running the code generator? That way we won't forget to fix them. Edit: one umbrella issue with all the asks is also fine.

**Comment by petar** [comment]
Created an issue: https://github.com/ipld/go-ipld-prime/issues/325
Can I get an approval from someone? Ready to merge?

**Review by mvdan — APPROVED** [review]
🛫

**Pull request merged**
